rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,272 @@
+
+To pass the attestation certificates to the Auth Server, they will be added to the login endpoints, including `/webali/ssh/certs`, `/webapi/mfa/login/finish`, `/webapi/<sso>/callback`.
+
+Additionally, the attestation will return useful information about the user's yubikey and the private key in question, including:
+ - The private key's PIN policy
+ - The private key's Touch policy
+ - The yubikey serial number
+
+#### Auth Service Config or Auth Preference
+
+To require yubikey login for all users, you can set the `require_yubikey_login` field in the Cluster's Auth Preference, either through the Auth Service's config file:
+
+```yaml
+auth_service:
+  ...
+  authentication:
+    ...
+    yubikey_login: 
+      required: true
+      pin_policy: always
+      touch_policy: always
+```
+
+Or with a dynamic Cluster Auth Preference object:
+
+```yaml
+kind: cluster_auth_preference
+version: v2
+metadata:
+  name: cluster-auth-preference
+spec:
+  yubikey_login: 
+    required: true
+    pin_policy: always
+    touch_policy: always
+```
+
+If `yubikey_login_required` is `false`, the policy fields will be ignored and attestation will be skipped.
+
+When `yubikey_login_required` is `true`, then the server will perform Attestation and verify the passed public key.
+
+If `pin_policy` or `touch_policy` is set, then the Auth server will check the [`Attestation`](https://pkg.go.dev/github.com/go-piv/piv-go@v1.9.0/piv#Attestation) object's `PINPolicy` or `TouchPolicy` fields. If the checked fields are at least as strict as the policies required, then the request will proceed.
+
+Examples:
+ - `pin_policy=once` + `PINPolicy=always` = success
+ - `touch_policy=always` + `TouchPolicy=always` = success
+ - `touch_policy=cached` + `TouchPolicy=never` = failure
+
+#### Role Options
+
+To make yubikey login required or optional on a per-role basis, you can set the same options in a Role.
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: role-name
+spec:
+  yubikey_login: 
+    required: true
+    pin_policy: once
+    touch_policy: cached
+```
+
+#### Combined yubikey requirements
+
+Each user's yubikey requirements will be determined by:
+ 1. Combining their role options to form the strictest setting
+ 2. Combining the strictest role options settings with the cluster auth preference, with role options taking precedence (if set).
+
+Note: A user's yubikey requirement settings cannot be easily determined before generating the yubikey private key and locking in its pin/touch policy settings, which occurs before logging in. Thus, it is the responsibility of the user to figure out their own pin/touch policy settings before logging in, or else they will get `access denied` errors.
```

Comment:
If I follow it correctly, we could finish the authn flow and do another hop to get their PIV-backed certificate, which allows us to fetch the settings beforehand. This changes things a bit, but has also the advantage that Yubi PIV certs would come from a single endpoint.

Reply:
Great idea, I've added it to the section `Retrieve user's PIV attestation requirements during login`. Let me know what you think or if you had a better solution in mind.

Reply:
I had a slightly different idea in mind, but GetPIVAssertionRequirements works too. Seems a bit simpler to implement too, which is a plus.